experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+    else:
+        # do we have a way to make this into a DTensor?
+        placements = get_placements(tensor)
+        if placements is None:
+            placements = default
+
+        # we don't have a way to make into DTensor, return either
+        # the original tensor if it's a materialized one, or a new example with
+        # zeros.
+        if placements is None:
+            return (
+                torch.zeros(tensor.shape, device=tensor.device)
+                if isinstance(tensor, FakeTensor) else tensor
+            )
+
+        if isinstance(tensor, FakeTensor):
+            shard = get_example_shard(tensor, mesh, default, fn=torch.zeros)
+            return DTensor.from_local(shard, device_mesh=mesh, placements=placements)
+        else:
+            # assume the input is a full real tensor, we distribute it.
+            return distribute_tensor(tensor, device_mesh=mesh, placements=placements)
+
+
+def materialize_module(
+    mod: nn.Module,
+    tensor_materializer: Callable[[FakeTensor], torch.Tensor] = _materialize_tensor,
+) -> Tuple[nn.Module, dict[torch.Tensor, torch.Tensor]]:
+    def to_real_tensor(m: nn.Module, key: str, t: T) -> Union[T, torch.Tensor, nn.Parameter]:
+        if isinstance(t, nn.Parameter):
+            return nn.Parameter(tensor_materializer(t.data, t))
+        else:
+            assert isinstance(t, torch.Tensor)
+            return tensor_materializer(t)
+
+    def materialize_this_module(m: nn.Module) -> None:
+        for key, param in m._parameters.items():
+            assert isinstance(param, nn.Parameter)
+            if True:  # if isinstance(param.data, FakeTensor):
+                dparam = to_real_tensor(m, key, param)
+                assert isinstance(dparam, nn.Parameter)
+                m.register_parameter(key, dparam)
+        for key, buffer in m._buffers.items():
+            if True:  # if isinstance(buffer, FakeTensor):
+                m._buffers[key] = to_real_tensor(m, key, buffer)
+
+    materialize_this_module(mod)
+    for name, submod in mod.named_modules():
+        materialize_this_module(submod)
+
+    return mod
+
+
+def get_example_shard(tensor, mesh, default=None, fn=torch.randn):
+    if not isinstance(tensor, Tensor):
+        return tensor
+    elif isinstance(tensor, DTensor):
+        return tensor.to_local()
+    elif hasattr(tensor, '_placements'):
+        inp_spec = DTensorSpec(mesh, tensor._placements, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    elif default is not None:
+        inp_spec = DTensorSpec(mesh, default, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    raise NotImplementedError(f'Dont know how to get a shard of tensor {tensor}')
+
+
+
+def make_inductor_fn(fn, example_inputs, default_shard_schema=None):
+    """
+    Use inductor to compile the function fn. 
+    Uses make_functional_fx to trace it passing example_inputs.
```

Comment:
are example_inputs always going to be Fake tensors?

Reply:
not necessraily although i didn't test that it still works if example inputs are not fake.

Reply:
I think there is a convention now where dynamo always fake-ifies inputs before passing them into the downstream compiler stack.  I'm assuming that since this function is analogous to the inductor backend to dynamo, the same should apply

Reply:
Makes sense -- I think there's no reason to not fakify example_inputs. However some of the inputs are stateful (such as models and optimizers) -- for these it's not clear if we should force-fakify them. As a matter of fact, we are actually materializing those inside this function. 


Ideally we want the materialize part to happen _after_ this compilation is done. The only reason I'm doing it beforehand right now is because SPMD expansion still doesn't work in Fake Mode.